### ISSUE TITLE
beforeSuite running on the main actor when not tagged to do so. WAS: XCTestExpectation not working anymore in v6 in beforeSuite

### ISSUE THREAD

**Issue by Kalzem** [Quick/Quick#1179]
### What did you do?

I have a XCTestExpectation for an API fetch that I can't mock properly (third party SDK) so I call it in beforeSuite.
Before v6, the following code worked. Now it keeps failing even though `fulfill()` is indeed called.

```swift
final class PurchaseManagerSpec: QuickSpec {
    override func spec() {
        beforeSuite {
            let exp = self.expectation(description: "Fetching IAPs")
            PurchaseManager.shared.fetchAllProducts { status in
                switch status {
                case .success: break
                case .error: print("some errors but I don't care")
                }
                exp.fulfill()
            }
            self.waitForExpectations(timeout: 20) // <- Always failing with the wrong message
        }
    ...
}
```

I also tried adding `@MainActor` to beforeSuite, it doesn't change anything.

### What did you expect to happen?

v6 should work like v5 as I don't see mention of XCTestExpectation with breaking changes in the changelog.

### What actually happened instead?

XCTestExpectation not working in beforeSuite anymore.

Also I noticed that Xcode is lost and doesn't understand which test failed.

The issue comes from PurchaseManagerSpec but the error I get from line `self.waitForExpectations(timeout: 20)` is

```
An_AnalyticsManager__can_send_all_achievements_events:(): 
  Asynchronous wait failed: Exceeded timeout of 20 seconds, with unfulfilled expectations: "Fetching IAPs".
```

which is another Spec class.

### Environment

 - Quick: 6.0.0
 - Nimble: 11.1.0
 - Xcode Version: 14.0.1
 - Swift Version: The one of Xcode 14.0.1
 - CocoaPods: 1.11.3

**Comment by younata** [comment]
I'm not able to replicate your error when doing this in a `beforeSuite`. Here's my test code:

```
final class FunctionalTests_XCTestExpectationSpec: QuickSpec {
    override func spec() {
        beforeSuite {
            let exp = self.expectation(description: "beforeSuite")
            Task {
                exp.fulfill()
            }
            self.wait(for: [exp], timeout: 10)
        }

        beforeEach {
            let exp = self.expectation(description: "beforeEach")
            Task {
                exp.fulfill()
            }
            self.wait(for: [exp], timeout: 10)
        }

        afterEach {
            let exp = self.expectation(description: "afterEach")
            Task {
                exp.fulfill()
            }
            self.wait(for: [exp], timeout: 10)
        }

        it("works with XCTestExpectation") {
            let exp = self.expectation(description: "it")
            Task {
                exp.fulfill()
            }
            self.wait(for: [exp], timeout: 10)
        }
    }
}
```

I am able to get the expectation in the `it` block to reliably time out (which doesn't occur if I use `DispatchQueue.global().async` instead of `Task` to run the closure that calls `exp.fulfill()`). But only the `it` block.

> Also I noticed that Xcode is lost and doesn't understand which test failed.

That's likely a result of Quick reporting the failure after the correct test has ended. Which is a result of the async migration. I thought I fixed those.

---

While digging in to this, I'm noticing that `beforeSuite`'s are run as part of `Example.run()` (similarly with `afterSuite`). Which is likely not a pattern we should continue with. We should probably hook into the `class func setup()` or `class func teardown()` methods for these. I don't expect this to fix this bug, but it's worth cleaning up.

---

I believe you that you're running in to this problem. I'm not able to replicate it myself.

As a temporary workaround, you could convert your `PurchaseManager.shared.fetchAllProducts` method into an async method (either directly or by using [`withCheckedThrowingContinuation`](https://developer.apple.com/documentation/swift/withcheckedthrowingcontinuation(function:_:))) and awaiting on that in your `beforeSuite`.

**Comment by Kalzem** [comment]
Thanks for answering @younata 

I'm trying to figure out what can cause the bug, but I'm not well versed with XCTestExpectation and Task.

I wonder if the bug is not thread related.

Basically what's under `PurchaseManager.shared.fetchAllProducts` is a native SKProductsRequest (Apple SDK fetching StoreKit products) which returns to a non-main thread which is then dispatched to the main thread with a `DispatchQueue.main.async(execute:)`.

I can't find any info on wether XCTestExpectation is thread-safe or not. But maybe there is a potential conflict with Quick v6 thread management.

I have moved my SK products fetching from beforeSuite to `override func setUp(completion: @escaping (Error?) -> Void)` and now it works :)

Thanks for pointing out this solution.

**Comment by younata** [comment]
> I'm trying to figure out what can cause the bug, but I'm not well versed with XCTestExpectation and Task.

> I wonder if the bug is not thread related.

> Basically what's under PurchaseManager.shared.fetchAllProducts is a native SKProductsRequest (Apple SDK fetching StoreKit products) which returns to a non-main thread which is then dispatched to the main thread with a DispatchQueue.main.async(execute:).

Oh. Actually. This is kinda thread related, and is appears to be due to 2 things:

1. XCTestExpectation does not work when it's running on the main actor, but awaiting for something else to run on the main actor/thread.

You can verify this with the following XCTest test.

```swift
class XCTestExpectationTests: XCTestCase {
    @MainActor func testXCTestExpectationAsync() async {
        let exp = self.expectation(description: "it")
        DispatchQueue.main.async {
            exp.fulfill()
        }
        self.wait(for: [exp], timeout: 10)
    }
}
```

(I'll file a ~radar~ feedback with this.)

However, this isn't the entire story. The below screenshot shows the 3 variants of awaiting on XCTestExpectation: async on the main actor, async off the main actor, and sync (on the main thread). This only fails when in the first case.

<img width="1179" alt="Screen Shot 2022-11-06 at 5 49 27 PM" src="https://user-images.githubusercontent.com/285321/200210569-42b9d003-dd22-4685-a4e2-49f75b509513.png">

However, unless you're manually tagging your `beforeSuite` with `@MainActor` (or tagging the entire enclosing class with `@MainActor`), then this should still be running in that "async, but off the main actor" case where XCTestExpectation still works. Which brings me to...

2. `beforeSuite` should not automatically be running on the main thread.

Basically, the following should pass:

```swift
class MySpec: QuickSpec {
    override func spec() {
        beforeSuite { expect(Thread.isMainThread).to(beFalse()) }

        // ...
    }
}
```

Which is apparently not the case. Wild. Again, as I said earlier, the only time that test should fail is if you either tag closure used in the beforeSuite with `@MainActor`, or you tag the entire class itself to run on the main actor.

I'm going to investigate this even further. Thank you!

**Comment by younata** [comment]
Quick 7 resolves this issue. Even if you use `AsyncSpec`, because the `beforeSuite`s are run on the main thread using `MainActor.run`, this won't trip this issue.

That said, your particular example will not compile in Quick 7. Due to the `spec` method being a class method now, you won't be able to use `self.waitForExpectations` as expected. The best way to approach this is to do something like this:

```swift
beforeSuite {
    guard let spec = currentSpec() else { return }
    let exp = spec.expectation(description: "Fetching IAPs")
    PurchaseManager.shared.fetchAllProducts { status in
        switch status {
        case .success: break
        case .error: print("some errors but I don't care")
        }
        exp.fulfill()
    }
    spec.waitForExpectations(timeout: 20) // <- Always failing with the wrong message
}
```

I'm going to close this now. Starting in Quick 7, `beforeSuite`s will always run in synchronous contexts, and on the main thread. This is the case for `QuickSpec` (i.e. Quick 5 and earlier), and `AsyncSpec` (which executes `beforeSuite` closures using `MainActor.run`, and I did verify that `XCTestExpectation`s will work as expected using this).

**Comment by sushant-here** [comment]
hi @younata any update on when we will be able to call async code from within `beforeSuite` blocks?

<img width="945" alt="image" src="https://github.com/user-attachments/assets/d3931883-01d2-4256-9399-bebbf26d7d85">

Planning to bring back with quick 8?
